Added option in preferences to use custom URL base for DOI generation

BJaroszkowski opened the pull request:
Added option in preferences to allow users to choose custom base of DOI address whenever trying to access article via DOI field as requested in #7337 . This is a draft pull request since placement in GUI and functionality are still up for discussion. 

![custom_doi](https://user-images.githubusercontent.com/54515221/109554347-63be3f80-7ad4-11eb-9f20-934da0adb88d.jpg)

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

calixtus commented on src/main/java/org/jabref/gui/fieldeditors/IdentifierEditor.java:
```diff
@@ -85,6 +90,11 @@ private void lookupIdentifier() {
     @FXML
     private void openExternalLink() {
-        viewModel.openExternalLink();
+        if (StandardField.DOI.equals(field) && preferences.getDOIPreferences().isUseCustom()) {
+            String baseURL = preferences.getDOIPreferences().getDefaultBaseURI();
+            viewModel.openDOIWithCustomBase(baseURL);
+        } else {
+            viewModel.openExternalLink();
+        }
```
Can you move this distinction to the viewModel?
(you probably have to inject the PreferencesService by constructor then too)

calixtus reviewed (changes requested):
Hi @BJaroszkowski , thanks for your contribution. Codewise the changes already very good.
One thing I personally would prefer would be to move this setting out of the GeneralTab to a new PreferenceTab ("Import" or likewise. Creating a new tab should be straightforward, there are plenty of examples). I'm sure, there are other settings in the preferences dialog, that can later be moved there too, to clean the dialog a bit up - but this would not be the focus of this PR.

calixtus commented:
Be aware, that there is still a localization test failing. Please add this string (`Use custom DOI base URI for article access`) to the english resource file. (see https://devdocs.jabref.org/getting-into-the-code/code-howtos#using-localization-correctly)

BJaroszkowski commented:
I moved the settings to new Preference tab and moved some of the logic to IdentifierEditorViewModel. The name of the tab is more of a placeholder for the moment. I don't think that "Import" is suitable since this particular setting deals with access to external resources.

I still need to update tests to reflect the changes to IdentifierEditor and IdentifierEditorViewModel.

calixtus commented:
Maybe it suits somewhere else in the preferences dialog? Maybe some more options can later be collected in this new tab?

BJaroszkowski commented:
I think the pull request in the current state is ready for review. The other options I could see being included in the new tab are:

- Entry Owner
- Merge in `Custom editor tabs` Tab
- Perhaps also `Custom import/export formats` although the tab could get slightly crowded

Siedlerchr commented on src/main/java/org/jabref/gui/fieldeditors/IdentifierEditorViewModel.java:
```diff
@@ -78,6 +94,18 @@ public void openExternalLink() {
     }
 
+    public void openDOIWithCustomBase(String baseURI) {
+        identifier.get().map(x -> (DOI) x).flatMap(x -> x.getExternalURIWithCustomBase(baseURI)).ifPresent(
```
```suggestion
        identifier.get().map(identifier-> (DOI) identifier).flatMap(DOI::getExternalURIWithCustomBase(baseURI)).ifPresent(
```

please no single char variable name like x, use a meaninfgul variable and inside the flatmap you should be able to use a static syntax like DOI::getExternalUriWithCustomBase

BJaroszkowski replied:
Correct me if I'm wrong but since flatMap requires Function and we have two input arguments (String baseUri and DOI) is it not impossible to implement the static syntax? IntelliJ seems to agree with that.

Siedlerchr replied:
Sorry, overlooked that, you are right

Siedlerchr commented on src/main/resources/l10n/JabRef_en.properties:
```diff
@@ -1,297 +1,159 @@
 Unable\ to\ monitor\ file\ changes.\ Please\ close\ files\ and\ processes\ and\ restart.\ You\ may\ encounter\ errors\ if\ you\ continue\ with\ this\ session.=Unable to monitor file changes. Please close files and processes and restart. You may encounter errors if you continue with this session.
 %0\ contains\ the\ regular\ expression\ <b>%1</b>=%0 contains the regular expression <b>%1</b>
-
```
Please revert the deletion of the blank lines, just add the new keys at the bottom

Siedlerchr commented on src/main/java/org/jabref/gui/preferences/customization/CustomizationTab.fxml:
```diff
@@ -0,0 +1,18 @@
+        <CheckBox fx:id="useCustomDOI" text="%Use custom DOI base URI for article access"/>
+        <TextField fx:id="useCustomDOIName" HBox.hgrow="ALWAYS"/>
+    </HBox>
+</fx:root>
```
missing newline at the end of the file

Siedlerchr reviewed (changes requested):
Thanks for your contribution in general I think this already looks good. Just a few minor remarks

calixtus reviewed (approved):
Thanks for your contribution. Looks good to me to.
One thing I'm not yet sure ist the wording for the customization tab, but this can also maybe changed in some other PR.
